doc/memos/rdm-draft-alamos-ieee802154-radio-hal.md

```diff
@@ -0,0 +1,908 @@
+          |              |                         |                                               
+          v              |                         v                                               
++---------------------+  |  +-----------------------------------------------+
+|                     |  |  |                                               |
+|    Device Driver    |  |  |                  Device Driver                |
+|                     |  |  |                                               |
++---------------------+  |  +-----------------------------------------------+
+```
+
+
+# 4. Architecture
+```
++-----------------------------------------------------------------------------+
+|                                                                             |
+|                               Upper layer                                   |
+|                                                                             |
++-----------------------------------------------------------------------------+
+      ^
+      |
+      |
+      |
+  Radio HAL API                                  +----------------------------+
+      |                         Radio HAL API    |                            |
+      |                   +----------------------|    Bottom-Half processor   |
+      |                   |     (Process IRQ)    |                            |
+      |                   |                      +----------------------------+
+      |                   |                                   ^
+      |                   |                                   |
+      v                   v                                  IRQ
++-----------------------------+                               |
+|                             |               HW independent  |
+|   IEEE802.15.4 Radio HAL    |------------------------------------------------
+|                             |               HW dependent    |
++-----------------------------+                               |
+                |                                             |
+       Device Specific API                                    | 
+                |                                             |
+                v                                             |
++-----------------------------+                               |
+|                             |                               |
+|       Device Driver         |-------------------------------+
+|       implementation        |
++-----------------------------+
+```
+
+As shown in the above figure, the IEEE802.15.4 Radio HAL is a central component
+that provides any upper layer a technology-dependent and unified access to the
+hardware-specific device driver, by implementing the Radio HAL API.  Since
+devices drivers do not depend on the Radio HAL, it is still possible to use a
+the raw driver of a specific device, for testing purposes or accessing device
+specific features.
```

Comment:
Actually I'm not strongly in favor of having the Radio HAL on top of the device driver, but I tried to mimic what we have in other devices (e.g `saul` on top of sensor devices). Besides testing, I agree that the use cases for a "raw device driver" are rare.

If people don't care about that, I'm fine with the approach in https://github.com/RIOT-OS/RIOT/pull/13943#discussion_r415260167

Reply:
Also note: When new drivers for "strange" transceivers (e.g. let's say the CC1101 or the PR for adding netdev integration of the nRF24L01+) get merged, it would be very helpful to already have a good testing application in place. If the Radio HAL is the only API and a common part of the HAL exists, a shared testing application will likely find it's way into RIOT. That would ease testing for new transceivers in a "one-of-a-kind" class a lot, as those wouldn't need their own test application. (Except for features unique to them, if implemented.)

Reply:
In general I'm OK with that approach. @miri64 @PeterKietzmann @benpicco @bergzand is it OK for you too?

Reply:
I don't seem to understand this point correctly. So far, a technology specific HAL for 802.15.4 devices has been proposed. @maribu proposes a common HAL component that includes "strange" transceivers. @jia200x is OK with that approach, in general. Isn't that contradictory?

Reply:
My idea is that the common HAL component is a subset of the 802.15.4 specific HAL. Thus, every device implementing the 802.15.4 Radio HAL would implicitly also be compliant with the common HAL.

So: The IEEE 802.15.4 HAL would be technology specific and technology aware, but it would have a generic subset. To me, this works fine and is not contradicting.

But I think what @jia200x above agreed to is allowing and encouraging device drivers to directly implement the functions of the Radio HAL, rather than writing an 802.15.4 driver with an internal API and a Radio HAL wrapper above, with the Radio HAL wrapper being the only user of the driver internal API.

(Please note: I'm not against sharing code between different IEEE 802.15.4 driver; quite the contrary. I just would like to prevent having e.g. a `ieee802154_driver_t::transmit()` function that contains a single line of code calling e.g. `at86rf2xx_transmit()`. This would add one function call of overhead without improving code structure or maintainability. So if there is indeed a reason to implement one function of the HAL as a layer on top of more low level driver function (e.g. avoiding code duplication), I'm also all in for that. But please let us avoid adding a layer for the sake of adding a layer.)